### PR TITLE
3.10 systemd shims

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
This is backport of #8840 for 3.10.

/cc @sjenning @derekwaynecarr

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by sjenning** [comment]
failures due to `docker-excluder` package not being found

**Comment by ravisantoshgudimetla** [comment]
@sjenning I am much more concerned about the error related to upgrade_major. I am continuously getting error related to `fatal: [ocp-master -> {{ lookup('env', 'RHCI_ocp_master_IP') }}]: FAILED! => {` in https://s3.amazonaws.com/aos-ci/ghprb/openshift/openshift-ansible/da31aa72bfd066d89215ac81dc556f2e51d00db3.2.1530221197289339610/output.log

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by ravisantoshgudimetla** [comment]
closing in favor of cherrypicking #8840 against 3.10.

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by sjenning** [comment]
Ok, one more idea.  Now that I look at the PR more closely. we are changing more than we need to.

We need to just lay down the units, not use them in other areas of the installer because that creates an ordering dependency.  So any place you replaced `command` with `service` basically, lets not do that.

File: roles/openshift_control_plane/tasks/registry_auth.yml [hunk 1]
```diff
@@ -17,6 +17,7 @@
   until: master_oreg_auth_credentials_create.rc == 0
   notify: restart master
 
+
```
Comment (sjenning) [inline roles/openshift_control_plane/tasks/registry_auth.yml]: why?

File: roles/openshift_control_plane/handlers/main.yml [hunk 2]
```diff
@@ -10,6 +10,13 @@
   until: result.rc == 0
   notify: verify API server
 
+- name: restart master controllers
+  command: "systemctl restart {{ openshift_service_type }}-master-controllers"
+  retries: 5
+  delay: 5
+  register: result
+  until: result.rc == 0
```
Comment (sjenning) [inline roles/openshift_control_plane/handlers/main.yml]: doesn't seem related to delivering unit files

File: roles/openshift_control_plane/tasks/static.yml [hunk 3]
```diff
@@ -61,14 +61,6 @@
   - apiserver.yaml
   - controller.yaml
 
-- name: Remove old files in /etc/sysconfig
-  file:
-    name: "{{ item }}"
-    state: absent
-  with_items:
-  - /etc/sysconfig/{{ openshift_service_type }}-master-api
-  - /etc/sysconfig/{{ openshift_service_type }}-master-controllers
```
Comment (sjenning) [inline roles/openshift_control_plane/tasks/static.yml]: why remove?

File: roles/openshift_control_plane/handlers/main.yml (hunk 2 above)
Comment (ravisantoshgudimetla) [reply roles/openshift_control_plane/handlers/main.yml]: But we don't call this handler from any where. You want this to be deleted as well?

File: roles/openshift_control_plane/tasks/systemd_units.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    dest: "/etc/systemd/system/{{ openshift_service_type }}-{{ item }}"
+    owner: root
+    group: root
+    mode: 600
+  with_items:
+  - master-api.service
+  - master-controllers.service
+  - master-etcd.service
+  notify:
+  - reload systemd units
```
Comment (sjenning) [inline roles/openshift_control_plane/tasks/systemd_units.yml]: need to add this handler back.  i think i thought it was a task before, not a handler.  last thing as long as nothing else changes.

**Comment by ravisantoshgudimetla** [comment]
GCP cluster provisioning is failing with `Message:  No package matching 'origin-docker-excluder-3.10**' found available, installed or updated:`

Upgrade tests are passing fine.

**Comment by sjenning** [comment]
@sdodson this PR has been reduced to only copying the new systemd units.  It does not change the playbooks to use those units in any way.  The CI test failures are unrelated.

This is our last blocker for 3.10.  Any way to BGB merge this if you approve?

**Comment by sjenning** [comment]
> There are a lot places these services are being used still

Until this PR, those units don't exist.  How is this not a problem currently?

**Comment by ravisantoshgudimetla** [comment]
@vrutkovs - Any other comments on this PR? If not, can you LGTM and approve?

**Comment by vrutkovs** [comment]
The upgrade works fine, but the units are launched for some reason and are failing:
```
# systemctl status origin-master-api
● origin-master-api.service - "origin Master API"
   Loaded: loaded (/etc/systemd/system/origin-master-api.service; enabled; vendor preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Mon 2018-07-02 16:43:37 UTC; 8s ago
  Process: 45986 ExecStop=/usr/local/bin/origin-master-api stop (code=exited, status=203/EXEC)
  Process: 45984 ExecStart=/usr/local/bin/origin-master-api start (code=exited, status=203/EXEC)
 Main PID: 45984 (code=exited, status=203/EXEC)
   Memory: 0B
   CGroup: /system.slice/origin-master-api.service
# systemctl status origin-master-etcd
● origin-master-etcd.service - "origin Master Etcd"
   Loaded: loaded (/etc/systemd/system/origin-master-etcd.service; disabled; vendor preset: disabled)
   Active: inactive (dead)
[root@ip-172-18-0-164 ec2-user]# systemctl status origin-master-controllers
● origin-master-controllers.service - "origin Master Controllers"
   Loaded: loaded (/etc/systemd/system/origin-master-controllers.service; enabled; vendor preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Mon 2018-07-02 16:43:38 UTC; 17s ago
  Process: 45992 ExecStop=/usr/local/bin/origin-master-controllers stop (code=exited, status=203/EXEC)
  Process: 45990 ExecStart=/usr/local/bin/origin-master-controllers start (code=exited, status=203/EXEC)
 Main PID: 45990 (code=exited, status=203/EXEC)
```
Not sure if this code is worth including

**Comment by ravisantoshgudimetla** [comment]
We could try disabling them by removing `reload systemd units` handler. I will make that change if you are fine.

**Comment by sjenning** [comment]
@ravisantoshgudimetla we need to set the mode of the shims to be at least owner executable.

**Comment by ravisantoshgudimetla** [comment]
@sjenning I can do that too. Just looking at the error, it seems we are getting a different one. We would have gotten a an permission error, I believe if it is because of 600 mode.

**Comment by ravisantoshgudimetla** [comment]
@sjenning I have made the permission change. Let's see if it goes through.

**Comment by ravisantoshgudimetla** [comment]
bot, retest this please

**Comment by sjenning** [comment]
@ravisantoshgudimetla you made the service units executable. services need to be 644.  the **shims** need to be 755.

File: roles/openshift_control_plane/templates/master-api.service.j2 [hunk 5]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description="{{ openshift_service_type }} Master API"
+Requires=docker.service
+After=docker.service
+
+[Service]
+Type=simple
+TimeoutStartSec=10m
+ExecStart="/usr/local/bin/{{ openshift_service_type }}-master-api start"
```
Comment (sjenning) [inline roles/openshift_control_plane/templates/master-api.service.j2]: remove quotes around ExecStart/ExecStop

File: roles/openshift_control_plane/templates/master-controllers.service.j2 [hunk 6]
```diff
@@ -0,0 +1,16 @@
+[Unit]
+Description="{{ openshift_service_type }} Master Controllers"
+Requires=docker.service
+After=docker.service
+
+[Service]
+Type=simple
+TimeoutStartSec=10m
+ExecStart="/usr/local/bin/{{ openshift_service_type }}-master-controllers start"
```
Comment (sjenning) [inline roles/openshift_control_plane/templates/master-controllers.service.j2]: here too

File: roles/openshift_control_plane/templates/master-etcd.service.j2 [hunk 7]
```diff
@@ -0,0 +1,16 @@
+[Unit]
+Description="{{ openshift_service_type }} Master Etcd"
+Requires=docker.service
+After=docker.service
+
+[Service]
+Type=simple
+TimeoutStartSec=10m
+ExecStart="/usr/local/bin/{{ openshift_service_type }}-master-etcd start"
```
Comment (sjenning) [inline roles/openshift_control_plane/templates/master-etcd.service.j2]: here too

**Comment by ravisantoshgudimetla** [comment]
/close

As discussed, we have decided not to go with systemd shims for 3.10 and 3.11
